File: src/Tether_02.jl
# Example two: Falling mass, attached to linear spring
using ModelingToolkit, OrdinaryDiffEq, LinearAlgebra

G_EARTH  = Float64[0.0, 0.0, -9.81]    # gravitational acceleration [m/s²]
L0::Float64 = -10.0                    # initial spring length      [m]

# defing the model, Z component upwards
@parameters mass=1.0 c_spring=50.0 damping=0.5 l0=L0
@variables t pos(t)[1:3] = [0.0, 0.0,  L0]
@variables   vel(t)[1:3] = [0.0, 0.0,  0.0] 
@variables   acc(t)[1:3] = G_EARTH
@variables unit_vector(t)[1:3]  = [0.0, 0.0, -sign(L0)]
@variables spring_force(t)[1:3] = [0.0, 0.0, 0.0]
@variables force(t) = 0.0 norm1(t) = abs(l0) spring_vel(t) = 0.0
D = Differential(t)

eqs = vcat(D.(pos)      ~ vel,
           D.(vel)      ~ acc,
           norm1        ~ norm(pos),
           unit_vector  ~ -pos/norm1,         # direction from point mass to origin
           spring_vel   ~ -unit_vector ⋅ vel,
           spring_force ~ (c_spring * (norm1 - abs(l0)) + damping * spring_vel) * unit_vector,
           acc          ~ G_EARTH + spring_force/mass)

@named sys = ODESystem(eqs, t)
simple_sys = structural_simplify(sys)

# running the simulation
duration = 10.0
dt = 0.02
tol = 1e-6
tspan = (0.0, duration)
ts    = 0:dt:duration

prob = ODEProblem(simple_sys, nothing, tspan)
@time sol = solve(prob, Rodas5(), dt=dt, abstol=tol, reltol=tol, saveat=ts)

# plotting the result
X = sol.t
POS_Z = stack(sol[pos], dims=1)[:,3]
VEL_Z = stack(sol[vel], dims=1)[:,3]

PyPlot.close()
line1, = plot(X, POS_Z, color="green", label="pos_z")
xlabel("time [s]")
ylabel("pos_z [m]")
PyPlot.grid(true)
twinx()
ylabel("vel_z [m/s]") 
tight_layout(rect=(0, 0, 0.98, 0.98))
line2, = plot(X, VEL_Z, color="red", label="vel_z")
legend(handles=[line1, line2]) 
nothing

File: examples/python/Tether_02.py
# -*- coding: utf-8 -*-
"""
Tutorial example simulating a falling mass, attached to a linear spring.

The model is written as a CasADi expression graph and integrated with SUNDIALS' CVODES,
which CasADi supplies with the exact Jacobian of the model.
"""
import os
import numpy as np
import matplotlib.pyplot as plt
import casadi as ca

G_EARTH  = np.array([0.0, 0.0, -9.81]) # gravitational acceleration
C_SPRING =  50.0                       # spring constant [N/m]
DAMPING  =  0.5                        # damping [Ns/m]
MASS     = 1.0                         # mass of the point-mass [kg]
L_0      = 10.0                        # initial spring length [m]

# Falling mass, attached to a linear spring anchored at the origin
# State vector y = mass.pos, mass.vel
def build_model():
    """ The mass and its spring as one CasADi expression graph. Returns the state vector
        `y`, its derivative `ydot` and the initial state `y0`. """
    y = ca.SX.sym('y', 6)
    pos, vel = y[0:3], y[3:6]
    norm1 = ca.norm_2(pos)
    unit_vector = -pos / norm1                     # direction from point mass to origin
    spring_vel = -ca.dot(unit_vector, vel)
    spring_force = (C_SPRING * (norm1 - abs(L_0)) + DAMPING * spring_vel) * unit_vector
    acc = G_EARTH + spring_force / MASS
    y0 = np.array([0.0, 0.0, -L_0, 0.0, 0.0, 0.0])   # pos, vel
    return y, ca.vertcat(vel, acc), y0

def run_example():
    y, ydot, y0 = build_model()
    time = np.linspace(0.0, 10.0, 501)
    sim = ca.integrator('sim', 'cvodes', {'x': y, 'ode': ydot}, 0.0, time[1:],
                        {'abstol': 1.0e-6, 'reltol': 1.0e-6})
    y_sol = np.column_stack([y0, np.array(sim(x0=y0)['xf'])]).T
    pos_z = y_sol[:, 2]
    vel_z = y_sol[:, 5]

    # saving the result for comparison with the Julia implementation
    os.makedirs("output", exist_ok=True)
    with open(os.path.join("output", "Tether_02_python.csv"), "w") as f:
        f.write("time,pos_z,vel_z\n")
        for t_i, pz_i, vz_i in zip(time, pos_z, vel_z):
            f.write(f"{t_i},{pz_i},{vz_i}\n")

    plt.gcf().canvas.manager.set_window_title("falling mass, linear spring")
    plt.ax1 = plt.subplot(111)
    plt.ax1.set_xlabel('time [s]')
    plt.plot(time, pos_z, color="green")
    plt.ax1.set_ylabel('pos_z [m]')
    plt.ax1.grid(True)
    plt.ax2 = plt.twinx()
    plt.ax2.set_ylabel('vel_z [m/s]')
    plt.plot(time, vel_z, color="red")

    if os.environ.get("TETHERS_BRIEF_PLOT") == "1":
        # show briefly and close automatically, e.g. when running the tests
        plt.pause(1)
        plt.close('all')
    else:
        plt.show()

if __name__ == '__main__':
    run_example()
